### ISSUE TITLE
tsserver getDocumentHighlights is really used in an editor?

### ISSUE THREAD

**Issue by angelozerr** [Microsoft/TypeScript#8217]
tsserver provides `getDocumentHighlights `command and I suppose that this comand is to support `syntax coloration (like JSX)` in the editor which consumes tsserver.

But I have not found an editor which consumes this command. VSCode seems using `TextMate`. So why VSCode doesn't use getDocumentHighlights ? Perhaps it's because of performance?  Is Visual Studio 2015 uses getDocumentHighlights for syntax coloration? 

In other words, is `getDocumentHighlights` command is recommanded for syntax coloration (with big file too)?

Many thanks for your answer.

**Comment by drake7707** [comment]
If I remember correctly the document highlights returns all occurrences of the token at the given position. For example if you place your cursor on a variable it will light up the same variable on other lines to show its usage.

**Comment by DanielRosenwasser** [comment]
It used to be called "get occurrences" which was a slightly more obvious name, but didn't reflect the actual usage. It's used for when the cursor is on top of a keyword of some sort and the editor will highlight several relevant occurrences. For some purposes, this is functionally equivalent to Find-All-References, but limited to a set of files. For others, it's just whatever we chose to be relevant. For instance:

Highlights on an identifier:

![image](https://cloud.githubusercontent.com/assets/972891/14679146/80d8e55a-06cd-11e6-9e55-215c2cff9311.png)

Highlights on the `return` keyword:

![image](https://cloud.githubusercontent.com/assets/972891/14679169/9828001a-06cd-11e6-8005-b21d08b1144a.png)

Highlights on an `if` or `else` keyword:

![image](https://cloud.githubusercontent.com/assets/972891/14679184/ac420c8a-06cd-11e6-9d62-7dbb4e7d5366.png)

Highlights on a `switch`/`case`/`default`/`break`.

![image](https://cloud.githubusercontent.com/assets/972891/14679273/f378ac6c-06cd-11e6-9f64-f6026b0776ce.png)

**Comment by angelozerr** [comment]
Many thanks for your answer!

@DanielRosenwasser the samples that you have attached works too with tsserver `occurrences` command. 

VSCode uses tsserver `occurrences` command (see https://github.com/Microsoft/vscode/blob/master/extensions/typescript/src/features/documentHighlightProvider.ts#L31) and never `getDocumentHighlights` command.

That's why I was curious to know what is the goal of `getDocumentHighlights?
